mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,207 @@
+//===- SYCLToGPU.cpp - SYCL to GPU Patterns -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert SYCL dialect to GPU dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/SYCLToGPU/SYCLToGPU.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "mlir/Support/LLVM.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+namespace {
+/// Returns the result of creating an operation to get a reference to an element
+/// of a sycl::id or sycl::range.
+Value createGetOp(OpBuilder &builder, Location loc, Type underlyingArrTy,
+                  Value res, Value index, ArrayAttr argumentTypes,
+                  FlatSymbolRefAttr functionName) {
+  return TypeSwitch<Type, Value>(res.getType())
+      .Case<IDType, RangeType>([&](auto arg) {
+        // `this` type
+        using ArgTy = decltype(arg);
+        // Operation type depending on ArgTy
+        using OpTy = std::conditional_t<std::is_same_v<ArgTy, IDType>,
+                                        SYCLIDGetOp, SYCLRangeGetOp>;
+        return builder.create<OpTy>(
+            loc, underlyingArrTy, res, index, argumentTypes, functionName,
+            functionName,
+            builder.getAttr<FlatSymbolRefAttr>(ArgTy::getMnemonic()));
+      });
+}
+
+/// Creates an operation of with name \p opName from the GPU dialect using the
+/// dimension \p i as an attribute.
+///
+/// \return The result of such operation.
+Value getDimension(OpBuilder &builder, Location loc, StringRef opName,
+                   StringRef dimensionAttrName, int64_t i) {
+  assert(0 <= i && i < gpu::GPUDialect::getNumWorkgroupDimensions() &&
+         "Invalid index value");
+  // The GPU dialect lacks a proper way to obtain this name.
+  auto *op = builder.create(
+      loc, builder.getStringAttr(opName), ValueRange{}, builder.getIndexType(),
+      builder.getNamedAttr(
+          dimensionAttrName,
+          builder.getAttr<gpu::DimensionAttr>(static_cast<gpu::Dimension>(i))));
+  assert(op->getNumResults() == 1 && "Invalid number of results");
+  return op->getResult(0);
```

Comment:
Can we use `getResult()` here ?

Reply:
Unfortunately, we cannot. `op` is of type `Operation *`, as I took a generic approach and the underlying operation could be any of the `gpu` dialect operations we're replacing our operations with. I'm checking the number of results is correct for safety, though.